janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/strategy/JanusGraphMultiQueryStrategy.java

```diff
@@ -0,0 +1,75 @@
+// Copyright 2021 JanusGraph Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package org.janusgraph.graphdb.tinkerpop.optimize.strategy;
+
+import org.apache.tinkerpop.gremlin.process.traversal.Traversal.Admin;
+import org.apache.tinkerpop.gremlin.process.traversal.TraversalStrategy;
+import org.apache.tinkerpop.gremlin.process.traversal.strategy.AbstractTraversalStrategy;
+import org.apache.tinkerpop.gremlin.process.traversal.strategy.decoration.SubgraphStrategy;
+import org.apache.tinkerpop.gremlin.process.traversal.util.TraversalHelper;
+import org.apache.tinkerpop.gremlin.structure.Graph;
+import org.janusgraph.graphdb.database.StandardJanusGraph;
+import org.janusgraph.graphdb.tinkerpop.optimize.JanusGraphTraversalUtil;
+import org.janusgraph.graphdb.tinkerpop.optimize.step.JanusGraphMultiQueryStep;
+import org.janusgraph.graphdb.tinkerpop.optimize.step.MultiQueriable;
+import org.janusgraph.graphdb.transaction.StandardJanusGraphTx;
+
+import java.util.Collections;
+import java.util.Set;
+
+/**
+ * @author Marko A. Rodriguez (https://markorodriguez.com)
+ * @author Matthias Broecheler (http://matthiasb.com)
+ */
+public class JanusGraphMultiQueryStrategy extends AbstractTraversalStrategy<TraversalStrategy.ProviderOptimizationStrategy> implements TraversalStrategy.ProviderOptimizationStrategy {
+
+    private static final Set<Class<? extends ProviderOptimizationStrategy>> PRIORS = Collections.singleton(JanusGraphLocalQueryOptimizerStrategy.class);
```

Comment:
You are asking good questions...
I think what would happen is that no `MultiQuerySteps` aka caches are inserted because the query does not contain any `JanusGraphVertexStep`s or `JanusGraphPropertyStep`s. That _should_ work, because if there are none of these steps, that means there's also no step which is capable of using multi queries anyway. Therefore, there's no point in introducing caches.

Reply:
One solution to not have dependent strategies would be to move the code directly into the `JanusGraphLocalQueryOptimizerStrategy`, of course. But that would bloat the other strategy of course.

Reply:
would it be possible to have a simple test for that, just to ensure it runs properly and correctly?

Reply:
I added a simple test which uses a traversal that covers two cases: One where a` JanusGraphMultiQueryStep` can not be inserted because the existing `VertexStep` is not converted to a `JanusGraphVertexStep`. In the other case, a `JanusGraphMultiQueryStep` is inserted in front of a MultiQuery compatible parent. The test ensures that the traversal is still executable and produces the same result.